src/passes/LLVMNontrappingFPToIntLowering.cpp

```diff
@@ -0,0 +1,166 @@
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+#include <limits>
+
+// By default LLVM emits nontrapping float-to-int instructions to implement its
+// fptoui/fptosi conversion instructions. This pass replaces these instructions
+// with code sequences which also implement LLVM's fptoui/fptosi, but which are
+// not semantically equivalent in wasm. This is because out-of-range inputs to
+// these instructions produce poison values. So we need only ensure that there
+// is no trap, but need not ensure any particular result.
+
+// For example, if a conversion is guarded by a range check in the source, LLVM
+// can move the conversion before the check (and instead guard the use of the
+// result, which may be poison). This is valid in LLVM and for the nontrapping
+// wasm fptoint instructions but not for the trapping conversions. The
+// transformation in this pass is valid only if the nontrapping conversions
+// in the wasm were generated from LLVM and implement LLVM's conversion
+// semantics.
+
+namespace wasm {
+struct LLVMNonTrappingFPToIntLowering
+  : public WalkerPass<PostWalker<LLVMNonTrappingFPToIntLowering>> {
+
```

Comment:
```suggestion

  bool isFunctionParallel() override { return true; }

  std::unique_ptr<Pass> create() override {
    return std::make_unique<LLVMNonTrappingFPToIntLowering>();
  }

```
Seems like this can run in parallel on functions? However, if we did that we'd need to do more work below, as `run()` is no longer the entry point. Maybe not worth it, given that these lowering passes are expected to run rarely.

Reply:
What's the entry point then? can you point to a good example?

Reply:
You'll need two things. First, the entry is now `doWalkFunction` (since all that happens is to walk functions, in function-parallel mode). E.g.

https://github.com/WebAssembly/binaryen/blob/245d9156d8bd52ce9bc50c554aa3a9cd42cd0e6b/src/passes/MemoryPacking.cpp#L501-L507

Second, you can't change the features of the module in a function parallel pass, so you'd need a nested parallel one. A good example of that is here:

https://github.com/WebAssembly/binaryen/blob/245d9156d8bd52ce9bc50c554aa3a9cd42cd0e6b/src/passes/StripEH.cpp

Reply:
What do you think about this? It may not be worth it, as the speed is likely fine even on a large wasm, and this is just for legacy support.

Reply:
Sure, converted to function-parallel pass.